File: src/Tree/DiameterOfBinaryTree/DiameterOfBinaryTree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        #Base case
        if root is None:
            return 0
        
        diameter = 0
        
        def dfs(root): 
            left_height = 0
            right_height = 0
            
            if root.left is not None:
                left_height = dfs(root.left) + 1
            
            if root.right is not None:
                right_height = dfs(root.right) + 1
        
            
            nonlocal diameter
            diameter = max(left_height + right_height, diameter)
            
            return max(left_height, right_height)
        
        dfs(root)
        
        return diameter
            

File: src/Tree/DiameterOfBinaryTree/DiameterOfBinaryTree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int max = Integer.MIN_VALUE;
    public int diameterOfBinaryTree(TreeNode root) {
        dfs(root);
        return max;
    }
    private int dfs(TreeNode node){       
        int left = 0, right = 0;
        
        if(node.left != null){
            left = dfs(node.left);
        }
        
        if(node.right != null){
            right = dfs(node.right);
        }
        max = Math.max(max, left + right);
        
        return Math.max(left, right) + 1;
    }
}
